### PR TITLE
Change compiler_replace to a class that stores the code objects directly

### PR DESCRIPTION
**Pull request by pfultz2**
This will enable retrieving the code object to do tuning in the future.

### CODE REVIEW

**Review by TedThemistokleous — APPROVED** [review]

File: src/targets/gpu/include/migraphx/gpu/compiler.hpp [hunk 1]
```diff
@@ -38,7 +38,34 @@ namespace gpu {
...
+    template <class F>
+    compiler_replace(const operation& op, F f)
+        : operators{op},
+          replace_fn([=](const compiler_replace& cr, module& m, instruction_ref ins) {
+              f(m, ins, cr.operators.front());
+          })
+    {
+    }
+
+    std::vector<operation> operators = {};
```
Comment (umangyadav) [inline src/targets/gpu/include/migraphx/gpu/compiler.hpp:56]: I don't see need to keep it vector right now.

File: src/targets/gpu/jit/mlir.cpp [hunk 2]
```diff
@@ -45,10 +45,10 @@ struct mlir_compiler : compiler<mlir_compiler>
 
     compiler_replace insert(code_object_op co) const
     {
-        return [co = std::move(co)](module& m, instruction_ref ins) {
-            auto mlir = insert_mlir(m, ins, co, ins->inputs());
-            m.replace_instruction(ins, mlir);
-        };
+        return {co, [](module& m, instruction_ref ins, const operation& op) {
```
Comment (umangyadav) [inline src/targets/gpu/jit/mlir.cpp:48]: Fix tidy

**Comment by umangyadav** [comment]
> This will enable retrieving the code object to do tuning in the future.

Does HIP API has tuning support or the CK or MLIR  ?

**Comment by pfultz2** [comment]
> Does HIP API has tuning support or the CK or MLIR ?

HIP API doesnt have any tuning API. And CK and MLIR require manual tuning.

**Review by umangyadav — APPROVED** [review]

**Comment by pfultz2** [comment]
@causten This can be merged.

**Pull request merged**